From ef68fbaf5bec30e396800df0f388ce945c338c1e Mon Sep 17 00:00:00 2001
From: Christophe Guillon <christophe.guillon@st.com>
Date: Thu, 11 Sep 2014 01:52:01 +0200
Subject: [PATCH] Add support for P pass-through permission

Add P permission, which lets gitolite pass through to the
git backend even if the repository exists.
It is enabled only ig GL_WILDREPOS is set in .gitolite.rc.

Tested on top of gitolite v2.2.1
---
 src/gitolite.pm     |  9 ++++++++-
 src/gl-auth-command | 12 +++++++++---
 src/gl-compile-conf |  4 +++-
 3 files changed, 20 insertions(+), 5 deletions(-)

diff --git a/src/gitolite.pm b/src/gitolite.pm
index 4d5f9e2..98e9fe0 100644
--- a/src/gitolite.pm
+++ b/src/gitolite.pm
@@ -729,6 +729,7 @@ sub parse_acl
         for my $u ('@all', "$gl_user - wild", @user_plus, keys %perm_cats) {
             my $du = $gl_user; $du = '@all' if $u eq '@all' or ($perm_cats{$u} || '') eq '@all';
             $repos{$dr}{C}{$du} = 1 if $repos{$r}{C}{$u};
+            $repos{$dr}{P}{$du} = 1 if $repos{$r}{P}{$u};
             $repos{$dr}{R}{$du} = 1 if $repos{$r}{R}{$u};
             $repos{$dr}{W}{$du} = 1 if $repos{$r}{W}{$u};
 
@@ -798,6 +799,7 @@ sub add_repo_conf
             $wild = parse_acl($repo, $ENV{GL_USER});
         }
 
+        my $p_perm = ( $repos{$repo}{P}{'@all'} || $repos{$repo}{P}{$ENV{GL_USER}} ? 'P' : '' ) if ($GL_WILDREPOS);
         if ($exists) {
             if ($creator and $wild) {
                 $creator = "($creator)";
@@ -808,7 +810,7 @@ sub add_repo_conf
             } else {
                 $creator = "<gitolite>";
             }
-        } else {
+        } elsif ($p_perm eq "P") {
             # repo didn't exist; C perms need to be filled in
             $perm = ( $repos{$repo}{C}{'@all'} ? ' @C' : ( $repos{$repo}{C}{$ENV{GL_USER}} ? ' =C' : '   ' )) if $GL_WILDREPOS;
             # if you didn't have perms to create it, delete the "convenience"
@@ -818,6 +820,11 @@ sub add_repo_conf
         }
         $perm .= perm_code( $repos{$repo}{R}{'@all'} || $repos{'@all'}{R}{'@all'}, $repos{'@all'}{R}{$ENV{GL_USER}}, $repos{$repo}{R}{$ENV{GL_USER}}, 'R' );
         $perm .= perm_code( $repos{$repo}{W}{'@all'} || $repos{'@all'}{W}{'@all'}, $repos{'@all'}{W}{$ENV{GL_USER}}, $repos{$repo}{W}{$ENV{GL_USER}}, 'W' );
+        $perm = $p_perm if ($p_perm eq "P" and
+                            ! $repos{$repo}{R}{'@all'} and
+                            ! $repos{$repo}{R}{'@all'} and
+                            ! $repos{$repo}{W}{$ENV{GL_USER}} and
+                            ! $repos{$repo}{W}{$ENV{GL_USER}});
         $perm =~ s/./ /g if check_deny_repo($repo);
 
         # set up for caching %repos
diff --git a/src/gl-auth-command b/src/gl-auth-command
index 851f614..c11057d 100755
--- a/src/gl-auth-command
+++ b/src/gl-auth-command
@@ -162,11 +162,11 @@ if ( $GL_ALL_READ_ALL and $verb =~ $R_COMMANDS and -d "$REPO_BASE/$repo.git") {
 # it was missing, and you have create perms, so create it
 new_wild_repo($repo, $user) if ($perm =~ /C/);
 
-die "$aa access for $repo DENIED to $user
-(Or there may be no repository at the given path. Did you spell it correctly?)\n" unless $perm =~ /$aa/;
+die "$aa access for $repo DENIED to $user (perm: $perm)
+(Or there may be no repository at the given path. Did you spell it correctly?)\n" unless ($perm =~ /$aa/ || $perm =~ /P/);
 
 # check if repo is write-enabled
-check_repo_write_enabled($repo) if $aa eq 'W';
+check_repo_write_enabled($repo) if ($aa eq 'W' && $perm !~ /P/);
 
 # run the pre-git hook if present (do this last, just before actually handing
 # off to git).  Force its output to go to STDERR so the git client does not
@@ -191,5 +191,11 @@ if ($ENV{REQUEST_URI}) {
 
 log_it();
 
+if ($perm =~ /P/ and $verb ne 'git-init') {
+    log_it("pass-through: $verb '$REPO_BASE/$repo.git'");
+    $verb =~ s/^git /git-/;
+    exec("$verb", "$REPO_BASE/$repo.git")
+}
+
 $repo = "'$REPO_BASE/$repo.git'";
 exec("git", "shell", "-c", "$verb $repo") unless $verb eq 'git-init';
diff --git a/src/gl-compile-conf b/src/gl-compile-conf
index 910a2a9..b22a293 100755
--- a/src/gl-compile-conf
+++ b/src/gl-compile-conf
@@ -163,13 +163,14 @@ sub parse_conf_line
         s/\bCREAT[EO]R\b/\$creator/g for @{ $repos_p };
     }
     # actual permission line
-    elsif ($line =~ /^(-|C|R|RW\+?(?:C?D?|D?C?)) (.* )?= (.+)/)
+    elsif ($line =~ /^(-|P|C|R|RW\+?(?:C?D?|D?C?)) (.* )?= (.+)/)
     {
         my $perms = $1;
         my @refs; @refs = split( ' ', $2 ) if $2;
         @refs = expand_list ( @refs );
         my @users = split ' ', $3;
         die "$ABRT \$GL_WILDREPOS is not set, you cant use 'C' in config\n" if $perms eq 'C' and not $GL_WILDREPOS;
+        die "$ABRT \$GL_WILDREPOS is not set, you cant use 'P' in config\n" if $perms eq 'P' and not $GL_WILDREPOS;
 
         # if no ref is given, this PERM applies to all refs
         @refs = qw(refs/.*) unless @refs;
@@ -215,6 +216,7 @@ sub parse_conf_line
 
                 # for 1st level check (see faq/tips doc)
                 $repos{$repo}{C}{$user} = 1, next if $perms eq 'C';
+                $repos{$repo}{P}{$user} = 1, next if $perms eq 'P';
                 $repos{$repo}{R}{$user} = 1 if $perms =~ /R/;
                 $repos{$repo}{W}{$user} = 1 if $perms =~ /W|D/;
 
-- 
1.8.4

